feat: Added documentation for camera manager together with a few fixes (#1955)
cognitedata/reveal: documentation/docs/extending/camera-manager.mdx

---
id: camera-manager
title: Custom CameraManager
description: Shows how to implement custom CameraManager to enable new camera manipulation features.
---

import useBaseUrl from '@docusaurus/useBaseUrl';

By default `Cognite3DViewer` uses `DefaultCameraManager` class to manage user interaction with the camera.
In certain cases it could be useful to customize behaviour of the camera:

- Need to provide custom `THREE.PerspectiveCamera` object to `Cognite3DViewer`
- Need to extend camera controls to any specific case that is not covered by current implementation

## Overview

To create a custom camera manager class, `CameraManager` interface from `@cognite/reveal` must be implemented
and provided to `Cognite3DViewer` on construction using the `cameraManager`-option. Interface looks like this:

```js

> export interface CameraManager {

Author: Sure, I pasted it because I remember we can't link to reference. Probably a good idea is to finally update docusaurus and check whether it will fix it.

> export interface CameraManager {

Author: So, I think we can leave it as is and have a reference after we updated docusaurus.

  /**
   * Returns the camera used for rendering in {@link Cognite3DViewer}.
   * Note that the camera will not be modified directly by Reveal.
   * Implementations must trigger the `cameraChange`-event whenever the
   * camera changes.
   */
  getCamera(): THREE.PerspectiveCamera;
  /**

>    * Set camera's state. Rotation and target can't be set at the same time. as they could conflict.

Author: I think so, added the description for that.

   * @param deltaTime Delta time since last update in seconds.
   * @param boundingBox Global bounding box of the model(s) and any custom objects added to the scene.
   */
  update(deltaTime: number, boundingBox: THREE.Box3): void;
  /**
   * @obvious
   */
  dispose(): void;
}

```

Main implementation specific functions are `setCameraState`, `fitCameraToBoundingBox` and `update`. 
Setting state shouldn't be possible when `rotation` and `target` are passed at the same time because they could conflict.
When implementing these functions you can use a helper class `CameraManagerHelper` that contains some useful methods.

## Example implementation

Here is an example implementation of a custom camera manager that utilizes standard ThreeJS `OrbitControls` for mouse movement:

```js
import * as THREE from 'three';
import { OrbitControls } from 'three/examples/jsm/controls/OrbitControls';

import { CameraManager, CameraManagerHelper, CameraState, CameraChangeDelegate } from '@cognite/reveal';

export class CustomCameraManager implements CameraManager {
    private _domElement: HTMLElement;
    private _camera: THREE.PerspectiveCamera;
    private _controls: OrbitControls;

>     private _cameraManagerHelper = new CameraManagerHelper();

Reviewer: Can we not make this a static class so we instead do `CameraManagerHelper.helpMe()`?

Author: Yeah, will do that.

        return {
            position: this._camera.position.clone(),
            target: this._controls.target.clone(),
            rotation: this._camera.quaternion.clone(),
        }
    }

    on(event: "cameraChange", callback: CameraChangeDelegate): void {
        this._cameraChangedListener.push(callback);
    }

    off(event: "cameraChange", callback: CameraChangeDelegate): void {
        const index  = this._cameraChangedListener.indexOf(callback);
        if (index !== -1) {
            this._cameraChangedListener.splice(index, 1);
        }
    }

    fitCameraToBoundingBox(boundingBox: THREE.Box3, duration?: number, radiusFactor?: number): void {
        const { position, target } = this._cameraManagerHelper
          .calculateCameraStateToFitBoundingBox(this._camera, boundingBox, radiusFactor);

        this.setCameraState({ position, target });
    }

    update(deltaTime: number, boundingBox: THREE.Box3): void {
        this._controls.update();
        this._cameraManagerHelper.updateCameraNearAndFar(this._camera, boundingBox);
    }

    dispose(): void {
        this._controls.dispose();
        this._cameraChangedListener.splice(0);
    }
}
```

> As you can see, implementation could be really simple. There are three things that you should really pay attention to:

Reviewer: ```suggestion
As you can see, implementation can be quite simple. There are three things that you should pay extra attention to:
```

> 1. `'cameraChange'` event must be fired every time camera position or rotation changes because reveal rendering pipeline
>  relies on that.

Reviewer: ```suggestion
1. `'cameraChange'` event must be fired every time camera position or rotation changes to trigger redraw in Reveal```
